### PR TITLE
Performance improvements on package generation

### PR DESCRIPTION
**Pull request by dfmedeiros**
Right now, when Solidus tries to create the packages on `Coordinator`, it iterates over each one of the active stock locations and skip those without the requested variants in stock.

Let's suppose we have 100 different stock locations (which is my case) and the current user has 5 items on his cart. Once it tries to create the packages it triggers 500 different queries to the database.

Here I try to address this issue, iterating over just the stock locations with the requested variants in stock.

### CODE REVIEW

**Comment by cbrunsdon** [comment]
@dfmedeiros this looks fantastic, really appreciated.

I'm slammed by BlackFriday/Monday stuff but will pull it down and give it a run early next week. On first glance everything looks great though.

**Comment by jhawthorn** [comment]
Looks good to me. :+1:

I bet we could go further here and select the stock items grouped by stock_location, and skip the N+1 on line 59, but this is an excellent step towards that.

**Comment by dfmedeiros** [comment]
@cbrunsdon - Thanks for the feedback!

@jhawthorn - Yep, good catch! But I didn't get it correctly, what do you mean by `select the stock items grouped by stock_location`?

My plan is to select the correct stock items by `variant_id` including the stock_location and the variant itself, so we could check inside the loop. I think this would prevent the N + 1 as well. Something like:

```ruby
def requested_stock_items
  Spree::StockItem.where(variant_id: unnalocated_variant_ids).includes(:variant, :stock_location)
end

def build_packages(packages = Array.new)
  requested_stock_items.each do |stock_item|
    units_for_location = unallocated_inventory_units.select { |unit| unit.variant.eql?(stock_item.variant) }
    packer = build_packer(stock_item.stock_location, units_for_location)
    packages += packer.packages
  end
  packages
end
```

What do you think about this approach? Do you recommend doing something else?
Thanks again.

**Comment by athal7** [comment]
:+1: on this and on @jhawthorn's suggestion

**Comment by dfmedeiros** [comment]
@jhawthorn - Forget about my last comment, it doesn't work as I expected, but, I didn't get your last comment to select stock items grouped by stock location. I don't see how this would do any benefit for the queries on line 59.

File: core/app/models/spree/stock/coordinator.rb [hunk 1]
```diff
@@ -70,6 +69,15 @@ def unallocated_inventory_units
         inventory_units - @preallocated_inventory_units
       end
 
+      def stock_locations_with_requested_variants
+        Spree::StockLocation.active.joins(:stock_items).
+          where(spree_stock_items: { variant_id: unnalocated_variant_ids }).uniq
+      end
+
+      def unnalocated_variant_ids
```
Comment (philbirt) [inline core/app/models/spree/stock/coordinator.rb]: unallocated*

**Comment by philbirt** [comment]
Solid :+1:

**Comment by jhawthorn** [comment]
@dfmedeiros no worries. We can worry about future improvements after addressing this. I think this is good to merge pending the spelling fix.

**Comment by dfmedeiros** [comment]
@jhawthorn - Great, just fixed the typo. Thanks.

**Comment by jhawthorn** [comment]
Thank you @dfmedeiros

**Pull request merged**